sdk/containerregistry/container-registry/src/containerRegistryChallengeHandler.ts

```diff
@@ -0,0 +1,220 @@
+  ChallengeCallbacks,
+  parseWWWAuthenticate
+} from "./bearerTokenChanllengeAuthenticationPolicy";
+import { AcrAccessToken, AcrRefreshToken, GeneratedClient } from "./generated";
+import * as Mappers from "./generated/models/mappers";
+import * as Parameters from "./generated/models/parameters";
+
+/**
+ * Handles challenge based authentication for Container Registry Service.
+ *```
+ *  The challenge-based authorization flow for ACR is illustrated in the following steps.
+ *  For example, GET /api/v1/acr/repositories translates into the following calls.
+ *      Step 1: GET /api/v1/acr/repositories
+ *  Return Header: 401: www-authenticate header - Bearer realm="{url}",service="{serviceName}",scope="{scope}",error="invalid_token"
+ *      Step 2: Retrieve the serviceName, scope from the WWW-Authenticate header.  (Parse the string.)
+ *      Step 3: POST /api/oauth2/exchange
+ *  Request Body : { service, scope, grant-type, aadToken with ARM scope }
+ *  Response Body: { acrRefreshToken }
+ *      Step 4: POST /api/oauth2/token
+ *  Request Body: { acrRefreshToken, scope, grant-type }
+ *  Response Body: { acrAccessToken }
+ *      Step 5: GET /api/v1/acr/repositories
+ *  Request Header: { Bearer acrTokenAccess }
+ *```
+ */
+export class ChallengeHandler implements ChallengeCallbacks {
+  constructor(
+    private authClient: GeneratedClient,
+    private options: GetTokenOptions & { claims?: string } = {}
+  ) {}
+
+  /**
+   * Updates  the authentication context based on the challenge.
+   */
+  async authenticateRequestOnChallenge(
+    challenge: string,
+    options: ChallengeCallbackOptions
+  ): Promise<boolean> {
+    // Once we're here, we've completed Step 1.
+
+    // Step 2: Parse challenge string to retrieve serviceName and scope, where scope is the ACR Scope
+    const { service, scope } = parseWWWAuthenticate(challenge);
+
+    if (!service) {
+      throw new Error("Failed to retrieve 'service' from challenge");
+    }
+
+    if (!scope) {
+      throw new Error("Failed to retrieve 'scope' from challenge");
+    }
+
+    // Step 3: Exchange AAD Access Token for ACR Refresh Token
+    const acrRefreshToken = await ExchangeAadAccessTokenForAcrRefreshTokenAsync(
+      this.authClient,
+      GetAuthorizationToken(options.request),
+      service,
+      this.options
+    );
+
+    if (!acrRefreshToken) {
+      throw new Error("Failed to retrieve 'service' from challenge");
+    }
+
+    // Step 4: Send in acrRefreshToken and get back acrAccessToken
+    const acrAccessToken = await ExchangeAcrRefreshTokenForAcrAccessTokenAsync(
+      this.authClient,
+      acrRefreshToken,
+      service,
+      scope,
+      this.options
+    );
+
+    // Step 5 - Authorize Request.  Note, we don't use SetAuthorizationHeader here, because it
+    // sets an AAD access token header, and at this point we're done with AAD and using an ACR access token.
+    options.request.headers.set("Authorization", `Bearer ${acrAccessToken}`);
```

Comment:
What you have here is perfect. I was wondering, would it help for future users if we changed `setAuthorizationHeader` to receive a string instead of an `AccessToken`? Since it does almost the same.

Reply:
yes I did experiment with that.  I think string is more general.

Reply:
That's awesome then! Let's take that update on the main core PR too 🌞 this is great early feedback.